org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java

```diff
@@ -1051,6 +1051,90 @@ public static String[] getVariableNameSuggestions(int variableKind, IJavaProject
 		return res.toArray(new String[res.size()]);
 	}
 
+	public static String[] getVariableNameSuggestions(int variableKind, IJavaProject project, ITypeBinding expectedType, Expression assignedExpression, Collection<String> excluded,
+			String identicalName, Collection<String> namesInSameMethod) {
+		LinkedHashSet<String> res= new LinkedHashSet<>(); // avoid duplicates but keep order
+
+		String typeName= null;
+		int dim= 0;
+		if (expectedType != null) {
+			expectedType= Bindings.normalizeTypeBinding(expectedType);
+			if (expectedType != null) {
+				if (expectedType.isArray()) {
+					dim= expectedType.getDimensions();
+					expectedType= expectedType.getElementType();
+				}
+				if (expectedType.isParameterizedType()) {
+					expectedType= expectedType.getTypeDeclaration();
+				}
+				typeName= expectedType.getName();
+			}
+		}
+
+		String recycledName= recycleNames(typeName, assignedExpression, excluded, identicalName, namesInSameMethod);
```

Comment:
I think it would make sense to not bother calling recycleNames() unless the expression is a MethodInvocation and add a comment (e.g. // if we have a method invocation, see if we can recycle some previously used variable name that is assigned the same expression.